src/main/java/com/commercetools/sync/producttypes/utils/AttributeDefinitionsUpdateActionUtils.java

```diff
@@ -153,33 +147,35 @@
                     }
                 ));
 
-        return oldAttributeDefinitions
-            .stream()
-            .map(oldAttributeDefinition -> {
-                final String oldAttributeDefinitionName = oldAttributeDefinition.getName();
-                final AttributeDefinitionDraft matchingNewAttributeDefinitionDraft =
+        final List<UpdateAction<ProductType>> updateActions = new ArrayList<>();
+
+        for (AttributeDefinition oldAttributeDefinition : oldAttributeDefinitions) {
+
+            final String oldAttributeDefinitionName = oldAttributeDefinition.getName();
+            final AttributeDefinitionDraft matchingNewAttributeDefinitionDraft =
                     newAttributesDefinitionsDraftsNameMap.get(oldAttributeDefinitionName);
-                return ofNullable(matchingNewAttributeDefinitionDraft)
-                    .map(attributeDefinitionDraft -> {
-                        // attribute type is required so if null we let commercetools to throw exception
-                        if (attributeDefinitionDraft.getAttributeType() != null) {
-                            if (haveSameAttributeType(oldAttributeDefinition, attributeDefinitionDraft)) {
-                                return buildActions(oldAttributeDefinition, attributeDefinitionDraft);
-                            } else {
-                                return Arrays.asList(
-                                    RemoveAttributeDefinition.of(oldAttributeDefinitionName),
-                                    AddAttributeDefinition.of(attributeDefinitionDraft)
-                                );
-                            }
-                        } else {
-                            return new ArrayList<UpdateAction<ProductType>>();
-                        }
-
-                    })
-                    .orElseGet(() -> singletonList(RemoveAttributeDefinition.of(oldAttributeDefinitionName)));
-            })
-            .flatMap(Collection::stream)
-            .collect(Collectors.toList());
+
+            if (matchingNewAttributeDefinitionDraft != null) {
+                if (matchingNewAttributeDefinitionDraft.getAttributeType() != null) {
+                    if (haveSameAttributeType(oldAttributeDefinition, matchingNewAttributeDefinitionDraft)) {
```

Comment:
what about the old `attributeDefinition` type? According to this implementation, If it's `null`, then incorrect behaviour will occur: If will go into the method `haveSameAttributeType` method and will return `false` signalling that the attribute types are different and thus generating update actions, which are actually not needed. However, it should throw an exception.

And also as discussed together yesterday the method: `haveSameAttributeType` could be better changed from accepting `AttributeDefinition` params to directly accepting `AttributeType` params. This will save you from `null` checking inside of it again..

Reply:
Also, are these utils not unit tested 😢 ?

Reply:
https://github.com/commercetools/commercetools-sync-java/pull/316/commits/36c17f87da027843e1b5337590b8f8e9ab7af7d7

Reply:
https://github.com/commercetools/commercetools-sync-java/pull/316/commits/ddb0a358cacd4bc9d1b6eba6b09beee2649c68b2